File: java/org.financer.client.javafx/src/main/java/org/financer/client/javafx/main/transactions/TransactionDialog.java
package org.financer.client.javafx.main.transactions;

import com.jfoenix.controls.JFXButton;
import com.jfoenix.controls.JFXComboBox;
import com.jfoenix.controls.JFXListView;
import com.jfoenix.controls.JFXTextField;
import javafx.application.Platform;
import javafx.scene.control.Label;
import javafx.scene.control.ListCell;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Region;
import javafx.scene.layout.VBox;
import org.controlsfx.glyphfont.FontAwesome;
import org.controlsfx.glyphfont.GlyphFont;
import org.controlsfx.glyphfont.GlyphFontRegistry;
import org.financer.client.connection.ServerRequestHandler;
import org.financer.client.format.I18N;
import org.financer.client.javafx.components.DatePicker;
import org.financer.client.javafx.components.DoubleField;
import org.financer.client.javafx.dialogs.FileChooserDialog;
import org.financer.client.javafx.dialogs.FinancerConfirmDialog;
import org.financer.client.javafx.dialogs.FinancerDialog;
import org.financer.client.javafx.dialogs.FinancerExceptionDialog;
import org.financer.client.javafx.format.JavaFXFormatter;
import org.financer.client.javafx.local.LocalStorageImpl;
import org.financer.shared.model.transactions.ContentAttachment;
import org.financer.util.collections.TreeUtil;
import org.financer.util.concurrency.FinancerExecutor;

import java.awt.*;
import java.io.*;
import java.nio.file.Files;
import java.time.LocalDate;
import java.util.*;
import java.util.concurrent.Executors;
import java.util.logging.Level;
import java.util.logging.Logger;

class TransactionDialog extends FinancerDialog<VariableTransaction> {

    private DoubleField amountField;
    private JFXComboBox<CategoryTree> categoryComboBox;
    private JFXTextField productField;
    private JFXTextField purposeField;
    private JFXTextField shopField;
    private DatePicker valueDateField;
    private JFXListView<Attachment> attachmentListView;
    private BaseCategory categories;

    private final Logger logger = Logger.getLogger("FinancerApplication");

    TransactionDialog(VariableTransaction transaction, BaseCategory categories) {
        super(transaction);

        this.categories = categories;
        this.prepareDialogContent();
        this.setDialogTitle(I18N.get("transaction"));
    }

    @Override
    protected Region getDialogContent() {
        HBox hBox = new HBox();
        hBox.setSpacing(30);

        GridPane gridPane = new GridPane();
        gridPane.setHgap(80);
        gridPane.setVgap(8);

        gridPane.add(new Label(I18N.get("amount")), 0, 0);
        this.amountField = new DoubleField();
        this.amountField.setId("amountTextField");
        gridPane.add(this.amountField, 1, 0);

        gridPane.add(new Label(I18N.get("category")), 0, 1);
        this.categoryComboBox = new JFXComboBox<>();
        this.categoryComboBox.setId("categoryComboBox");
        this.categoryComboBox.setPlaceholder(new Label(I18N.get("selectCategory")));

        gridPane.add(this.categoryComboBox, 1, 1);

        gridPane.add(new Label(I18N.get("product")), 0, 2);
        this.productField = new JFXTextField();
        this.productField.setId("productTextField");
        gridPane.add(this.productField, 1, 2);

        gridPane.add(new Label(I18N.get("purpose")), 0, 3);
        this.purposeField = new JFXTextField();
        this.purposeField.setId("purposeTextField");
        gridPane.add(purposeField, 1, 3);

        gridPane.add(new Label(I18N.get("shop")), 0, 4);
        this.shopField = new JFXTextField();
        this.shopField.setId("shopTextField");
        gridPane.add(this.shopField, 1, 4);

        gridPane.add(new Label(I18N.get("valueDate")), 0, 5);
        this.valueDateField = new DatePicker(new JavaFXFormatter(LocalStorageImpl.getInstance()));
        this.valueDateField.setValue(LocalDate.now());
        this.valueDateField.setId("valueDatePicker");
        gridPane.add(this.valueDateField, 1, 5);

        hBox.getChildren().add(gridPane);

        if (this.getValue() != null) {
            VBox attachmentsContainer = new VBox();
            attachmentsContainer.setSpacing(10);
            attachmentsContainer.setPrefHeight(200);
            attachmentsContainer.getChildren().add(new Label(I18N.get("attachments")));

            GlyphFont fontAwesome = GlyphFontRegistry.font("FontAwesome");
            JFXButton uploadAttachmentBtn = new JFXButton(I18N.get("upload"), fontAwesome.create(FontAwesome.Glyph.PLUS));
            uploadAttachmentBtn.setId("uploadAttachmentBtn");
            JFXButton openFileBtn = new JFXButton(I18N.get("openFile"), fontAwesome.create(FontAwesome.Glyph.FOLDER_OPEN));
            JFXButton deleteAttachmentBtn = new JFXButton(I18N.get("delete"), fontAwesome.create(FontAwesome.Glyph.TRASH));
            deleteAttachmentBtn.setId("deleteAttachmentBtn");

            uploadAttachmentBtn.setOnAction(event -> {
                FileChooserDialog dialog = new FileChooserDialog();
                dialog.setOnConfirm(this::onUploadAttachment);
            });

            openFileBtn.setOnAction(event -> onOpenAttachment());

            deleteAttachmentBtn.setOnAction(event -> onDeleteAttachment());

            HBox toolBox = new HBox();
            toolBox.setSpacing(8);
            toolBox.getChildren().add(uploadAttachmentBtn);
            toolBox.getChildren().add(openFileBtn);
            toolBox.getChildren().add(deleteAttachmentBtn);

            this.attachmentListView = new JFXListView<>();
            this.attachmentListView.setCellFactory(param -> new ListCell<>() {
                @Override
                protected void updateItem(Attachment item, boolean empty) {
                    super.updateItem(item, empty);

                    if (item == null || empty) {
                        setGraphic(null);
                    } else {
                        setText(item.getName());
                    }
                }
            });

            attachmentsContainer.getChildren().add(toolBox);
            attachmentsContainer.getChildren().add(this.attachmentListView);

            hBox.getChildren().add(attachmentsContainer);
        }

        return hBox;
    }

    @Override
    protected double getDialogWidth() {
        return this.getValue() == null ? 400 : 750;
    }

    @Override
    protected void prepareDialogContent() {
        categories.traverse(treeItem -> {
            if (!treeItem.isRoot() && (treeItem.getValue().getCategoryClass() == BaseCategory.CategoryClass.VARIABLE_EXPENSES ||
                    treeItem.getValue().getCategoryClass() == BaseCategory.CategoryClass.VARIABLE_REVENUE)) {
                categoryComboBox.getItems().add((CategoryTree) treeItem);
            }
        });
        this.categoryComboBox.setCellFactory(param -> new ListCell<>() {
            @Override
            protected void updateItem(CategoryTree item, boolean empty) {
                super.updateItem(item, empty);
                if (!empty) {
                    setText(new JavaFXFormatter(LocalStorageImpl.getInstance()).formatCategoryName(item.getValue()));
                } else {
                    setText(null);
                }
            }
        });

        if (this.getValue() != null) {
            this.amountField.setText(String.valueOf(this.getValue().getAmount()));
            this.categoryComboBox.getSelectionModel().select((CategoryTree) TreeUtil.getByValue(this.categories,
                    this.getValue().getCategoryTree(), Comparator.comparingInt(Category::getId)));
            this.productField.setText(this.getValue().getProduct());
            this.purposeField.setText(this.getValue().getPurpose());
            this.shopField.setText(this.getValue().getShop());
            this.valueDateField.setValue(this.getValue().getValueDate());

            if (this.getValue().getAttachments() != null) {
                this.attachmentListView.getItems().addAll(new ArrayList<>(this.getValue().getAttachments()));
            }
        }
    }

    @Override
    protected boolean checkConsistency() {
        boolean result = true;

        if (this.categoryComboBox.getSelectionModel().getSelectedItem() == null) {
            setErrorMessage(I18N.get("selectCategory"));
            result = false;
        }

        if (Double.valueOf(this.amountField.getText()).equals(0.0)) {
            setErrorMessage(I18N.get("selectValidAmount"));
            result = false;
        }

        return result;
    }

    @Override
    protected void onConfirm() {
        if (this.getValue() == null) {
            this.setValue(new VariableTransaction(0,
                    Double.parseDouble(this.amountField.getText()),
                    this.valueDateField.getValue(),
                    this.categoryComboBox.getSelectionModel().getSelectedItem(),
                    this.productField.getText(),
                    this.purposeField.getText(),
                    this.shopField.getText()));
            this.getValue().getCategoryTree().getTransactions().add(this.getValue());
        } else {
            this.getValue().setAmount(Double.parseDouble(this.amountField.getText()));
            this.getValue().setCategoryTree(this.categoryComboBox.getSelectionModel().getSelectedItem());
            this.getValue().setProduct(this.productField.getText());
            this.getValue().setPurpose(this.purposeField.getText());
            this.getValue().setValueDate(this.valueDateField.getValue());
            this.getValue().setShop(this.shopField.getText());
        }

        super.onConfirm();
    }

    private void onUploadAttachment(File attachmentFile) {
        if (attachmentFile != null) {
            byte[] attachmentContent = new byte[(int) attachmentFile.length()];
            try (BufferedInputStream bufferedReader = new BufferedInputStream(new FileInputStream(attachmentFile))) {
                if (bufferedReader.read(attachmentContent, 0, attachmentContent.length) != -1) {
                    Map<String, Serializable> parameters = new HashMap<>();
                    parameters.put("transaction", this.getValue());
                    parameters.put("attachment", new ContentAttachment(0, this.getValue(),
                            attachmentFile.getName(), LocalDate.now(), attachmentContent));
                    Executors.newCachedThreadPool().execute(new ServerRequestHandler((User) LocalStorageImpl.getInstance().readObject("user"),
                            "uploadTransactionAttachment", parameters, result -> {
                        attachmentListView.getItems().add((Attachment) result.getResult());
                        if (getValue().getAttachments() == null) {
                            getValue().setAttachments(new HashSet<>());
                        }
                        getValue().getAttachments().add((Attachment) result.getResult());
                    }));
                }
            } catch (IOException e) {
                logger.log(Level.SEVERE, e.getMessage(), e);
            }
        }
    }

    private void onOpenAttachment() {
        File file = new File(LocalStorageImpl.LocalStorageFile.TRANSACTIONS.getFile().getParent() +
                "/transactions/" + this.getValue().getId() + "/attachments/" +
                this.attachmentListView.getSelectionModel().getSelectedItem().getName());

        if (file.exists()) {
            try {
                Desktop.getDesktop().open(file);
            } catch (IOException e) {
                new FinancerExceptionDialog("Financer", e);
            }
        } else {
            if (!file.getParentFile().exists()) {
                file.getParentFile().mkdirs();
            }

            Map<String, Serializable> parameters = new HashMap<>();
            parameters.put("attachmentId", this.attachmentListView.getSelectionModel().getSelectedItem().getId());

            FinancerExecutor.getExecutor().execute(new ServerRequestHandler(
                    (User) LocalStorageImpl.getInstance().readObject("user"),
                    "getAttachment", parameters,
                    result -> {
                        try (FileOutputStream fileOutputStream = new FileOutputStream(file)) {
                            fileOutputStream.write(((ContentAttachment) result.getResult()).getContent());
                            Desktop.getDesktop().open(file);
                        } catch (IOException e) {
                            logger.log(Level.SEVERE, e.getMessage(), e);
                        }
                    }));
        }
    }

    private void onDeleteAttachment() {
        FinancerConfirmDialog confirmDialog = new FinancerConfirmDialog(I18N.get("confirmDeleteAttachment"));
        confirmDialog.setOnConfirm(result -> {
            Attachment attachment = this.attachmentListView.getSelectionModel().getSelectedItem();

            Map<String, Serializable> parameters = new HashMap<>();
            parameters.put("attachmentId", attachment.getId());

            FinancerExecutor.getExecutor().execute(new ServerRequestHandler(
                    (User) LocalStorageImpl.getInstance().readObject("user"), "deleteAttachment",
                    parameters, result1 -> {
                getValue().getAttachments().remove(attachment);
                File file = new File(LocalStorageImpl.LocalStorageFile.TRANSACTIONS.getFile().getParent() +
                        "/transactions/" + attachmentListView.getSelectionModel().getSelectedItem().getTransaction().getId() +
                        "/attachments/" + attachmentListView.getSelectionModel().getSelectedItem().getName());
                try {
                    if (file.exists()) {
                        Files.delete(file.toPath());
                    }
                    Platform.runLater(() -> attachmentListView.getItems().remove(attachmentListView.getSelectionModel().getSelectedItem()));
                } catch (IOException e) {
                    logger.log(Level.SEVERE, e.getMessage(), e);
                    Platform.runLater(() -> new FinancerExceptionDialog("Financer", new IOException("File could not be deleted")));
                }
            }));
        });
    }
}

File: java/org.financer.client.javafx/src/main/kotlin/org/financer/client/javafx/main/transactions/TransactionDialog.kt
package org.financer.client.javafx.main.transactions

import com.jfoenix.controls.JFXButton
import com.jfoenix.controls.JFXComboBox
import com.jfoenix.controls.JFXListView
import com.jfoenix.controls.JFXTextField
import javafx.application.Platform
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.scene.control.Label
import javafx.scene.control.ListCell
import javafx.scene.layout.GridPane
import javafx.scene.layout.HBox
import javafx.scene.layout.Region
import javafx.scene.layout.VBox
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.controlsfx.glyphfont.FontAwesome
import org.controlsfx.glyphfont.GlyphFontRegistry
import org.financer.client.domain.api.RestApi
import org.financer.client.domain.api.RestApiImpl
import org.financer.client.domain.model.category.Category
import org.financer.client.domain.model.category.CategoryRoot
import org.financer.client.domain.model.transaction.Attachment
import org.financer.client.domain.model.transaction.VariableTransaction
import org.financer.client.format.I18N.get
import org.financer.client.javafx.components.DatePicker
import org.financer.client.javafx.components.DoubleField
import org.financer.client.javafx.dialogs.FileChooserDialog
import org.financer.client.javafx.dialogs.FinancerConfirmDialog
import org.financer.client.javafx.dialogs.FinancerDialog
import org.financer.client.javafx.dialogs.FinancerExceptionDialog
import org.financer.client.javafx.format.JavaFXFormatter
import org.financer.client.javafx.local.LocalStorageImpl
import org.financer.shared.domain.model.value.objects.ValueDate
import org.financer.util.collections.Action
import org.financer.util.collections.Tree
import org.financer.util.collections.TreeUtil
import java.awt.Desktop
import java.io.*
import java.nio.file.Files
import java.time.LocalDate
import java.util.*
import java.util.logging.Level
import java.util.logging.Logger

internal class TransactionDialog(transaction: VariableTransaction?, private val categories: CategoryRoot?) : FinancerDialog<VariableTransaction?>(transaction) {
    private var amountField: DoubleField? = null
    private var categoryComboBox: JFXComboBox<Category>? = null
    private var productField: JFXTextField? = null
    private var descriptionField: JFXTextField? = null
    private var vendorField: JFXTextField? = null
    private var valueDateField: DatePicker? = null
    private var attachmentListView: JFXListView<Attachment>? = null
    private val restApi: RestApi = RestApiImpl()
    private val logger = Logger.getLogger("FinancerApplication")
    override val dialogContent: Region
        get() {
            val hBox = HBox()
            hBox.spacing = 30.0
            val gridPane = GridPane()
            gridPane.hgap = 80.0
            gridPane.vgap = 8.0
            gridPane.add(Label(get("amount")), 0, 0)
            amountField = DoubleField()
            amountField!!.id = "amountTextField"
            gridPane.add(amountField, 1, 0)
            gridPane.add(Label(get("category")), 0, 1)
            categoryComboBox = JFXComboBox()
            categoryComboBox!!.id = "categoryComboBox"
            categoryComboBox!!.placeholder = Label(get("selectCategory"))
            gridPane.add(categoryComboBox, 1, 1)
            gridPane.add(Label(get("product")), 0, 2)
            productField = JFXTextField()
            productField!!.id = "productTextField"
            gridPane.add(productField, 1, 2)
            gridPane.add(Label(get("purpose")), 0, 3)
            descriptionField = JFXTextField()
            descriptionField!!.id = "purposeTextField"
            gridPane.add(descriptionField, 1, 3)
            gridPane.add(Label(get("shop")), 0, 4)
            vendorField = JFXTextField()
            vendorField!!.id = "shopTextField"
            gridPane.add(vendorField, 1, 4)
            gridPane.add(Label(get("valueDate")), 0, 5)
            valueDateField = DatePicker(JavaFXFormatter(LocalStorageImpl.getInstance()))
            valueDateField!!.value = LocalDate.now()
            valueDateField!!.id = "valueDatePicker"
            gridPane.add(valueDateField, 1, 5)
            hBox.children.add(gridPane)
            if (value != null) {
                val attachmentsContainer = VBox()
                attachmentsContainer.spacing = 10.0
                attachmentsContainer.prefHeight = 200.0
                attachmentsContainer.children.add(Label(get("attachments")))
                val fontAwesome = GlyphFontRegistry.font("FontAwesome")
                val uploadAttachmentBtn = JFXButton(get("upload"), fontAwesome.create(FontAwesome.Glyph.PLUS))
                uploadAttachmentBtn.id = "uploadAttachmentBtn"
                val openFileBtn = JFXButton(get("openFile"), fontAwesome.create(FontAwesome.Glyph.FOLDER_OPEN))
                val deleteAttachmentBtn = JFXButton(get("delete"), fontAwesome.create(FontAwesome.Glyph.TRASH))
                deleteAttachmentBtn.id = "deleteAttachmentBtn"
                uploadAttachmentBtn.onAction = EventHandler { event: ActionEvent? ->
                    val dialog = FileChooserDialog()
                    dialog.setOnConfirm(Action { attachmentFile: File? -> onUploadAttachment(attachmentFile) })
                }
                openFileBtn.onAction = EventHandler { event: ActionEvent? -> onOpenAttachment() }
                deleteAttachmentBtn.onAction = EventHandler { event: ActionEvent? -> onDeleteAttachment() }
                val toolBox = HBox()
                toolBox.spacing = 8.0
                toolBox.children.add(uploadAttachmentBtn)
                toolBox.children.add(openFileBtn)
                toolBox.children.add(deleteAttachmentBtn)
                attachmentListView = JFXListView()
                attachmentListView!!.setCellFactory {
                    object : ListCell<Attachment>() {
                        override fun updateItem(item: Attachment, empty: Boolean) {
                            super.updateItem(item, empty)
                            if (empty) {
                                graphic = null
                            } else {
                                text = item.name
                            }
                        }
                    }
                }
                attachmentsContainer.children.add(toolBox)
                attachmentsContainer.children.add(attachmentListView)
                hBox.children.add(attachmentsContainer)
            }
            return hBox
        }

    override val dialogWidth: Double
        get() = (if (value == null) 400 else 750).toDouble()

    override fun prepareDialogContent() {
        categories!!.traverse { treeItem: Tree ->
            val category = treeItem as Category
            if (!category.isRoot && !category.categoryClass!!.isFixed) {
                categoryComboBox!!.items.add(category)
            }
        }
        categoryComboBox!!.setCellFactory {
            object : ListCell<Category>() {
                override fun updateItem(item: Category, empty: Boolean) {
                    super.updateItem(item, empty)
                    text = if (!empty) {
                        JavaFXFormatter(LocalStorageImpl.getInstance()).format(item)
                    } else {
                        null
                    }
                }
            }
        }
        if (value != null) {
            amountField!!.text = value!!.amount.toString()
            categoryComboBox!!.selectionModel.select(TreeUtil.getByValue(categories,
                    value!!.category, compareBy { t: Category -> t.id }) as Category)
            //            this.productField.setText(this.getValue().getDe());
            descriptionField!!.text = value!!.description
            vendorField!!.text = value!!.vendor
            valueDateField!!.value = value!!.valueDate!!.date
            attachmentListView!!.items.addAll(ArrayList(value!!.getAttachments()))
        }
    }

    override fun checkConsistency(): Boolean {
        var result = true
        if (categoryComboBox!!.selectionModel.selectedItem == null) {
            errorMessage = get("selectCategory")
            result = false
        }
        if (java.lang.Double.valueOf(amountField!!.text) == 0.0) {
            errorMessage = get("selectValidAmount")
            result = false
        }
        return result
    }

    override fun onConfirm() {
        if (value == null) {
            value = VariableTransaction()
                    .setId(0) //                    .setAmount(Double.parseDouble(this.amountField.getText()))
                    .setValueDate(ValueDate(valueDateField!!.value))
                    .setCategory(categoryComboBox!!.selectionModel.selectedItem)
                    .setDescription(descriptionField!!.text)
                    .setVendor(vendorField!!.text)
            value!!.category!!.transactions.add(value!!)
        } else {
//            this.getValue().setAmount(new Amount(Double.parseDouble(this.amountField.getText()))=;
            value!!.setCategory(categoryComboBox!!.selectionModel.selectedItem)
            value!!.setValueDate(ValueDate(valueDateField!!.value))
            value!!.setDescription(descriptionField!!.text)
            value!!.setVendor(vendorField!!.text)
        }
        super.onConfirm()
    }

    private fun onUploadAttachment(attachmentFile: File?) {
        if (attachmentFile != null) {
            val attachmentContent = ByteArray(attachmentFile.length().toInt())
            try {
                BufferedInputStream(FileInputStream(attachmentFile)).use { bufferedReader ->
                    if (bufferedReader.read(attachmentContent, 0, attachmentContent.size) != -1) {
                        val attachment = Attachment()
                                .setName(attachmentFile.name)
                                .setUploadDate(LocalDate.now())
                                .setContent(attachmentContent)
                        CoroutineScope(Dispatchers.IO).launch {
                            val result = restApi.createAttachment(value!!.id, attachment)
                            if (result != null) {
                                attachmentListView!!.items.add(result)
                                value!!.getAttachments().add(result)
                            }
                        }
                    }
                }
            } catch (e: IOException) {
                logger.log(Level.SEVERE, e.message, e)
            }
        }
    }

    private fun onOpenAttachment() {
        val file = File(LocalStorageImpl.LocalStorageFile.TRANSACTIONS.file.parent +
                "/transactions/" + value!!.id + "/attachments/" +
                attachmentListView!!.selectionModel.selectedItem.name)
        if (file.exists()) {
            try {
                Desktop.getDesktop().open(file)
            } catch (e: IOException) {
                FinancerExceptionDialog("Financer", e)
            }
        } else {
            if (!file.parentFile.exists()) {
                file.parentFile.mkdirs()
            }
            val parameters: MutableMap<String, Serializable> = HashMap()
            parameters["attachmentId"] = attachmentListView!!.selectionModel.selectedItem.id
            CoroutineScope(Dispatchers.IO).launch {
                val result = restApi.getAttachment(value!!.id, attachmentListView!!.selectionModel.selectedItem.id)
                if (result != null) {
                    try {
                        FileOutputStream(file).use { fileOutputStream ->
                            fileOutputStream.write(result.content)
                            Desktop.getDesktop().open(file)
                        }
                    } catch (e: IOException) {
                        logger.log(Level.SEVERE, e.message, e)
                    }
                }
            }
        }
    }

    private fun onDeleteAttachment() {
        val confirmDialog = FinancerConfirmDialog(get("confirmDeleteAttachment"))
        confirmDialog.setOnConfirm(Action { result: Boolean? ->
            val attachment = attachmentListView!!.selectionModel.selectedItem
            CoroutineScope(Dispatchers.IO).launch {
                restApi.deleteAttachment(value!!.id, attachment.id)
                value!!.getAttachments().remove(attachment)
                val file = File(LocalStorageImpl.LocalStorageFile.TRANSACTIONS.file.parent +
                        "/transactions/" + attachmentListView!!.selectionModel.selectedItem.transaction!!.id +
                        "/attachments/" + attachmentListView!!.selectionModel.selectedItem.name)
                try {
                    if (file.exists()) {
                        Files.delete(file.toPath())
                    }
                    Platform.runLater { attachmentListView!!.items.remove(attachmentListView!!.selectionModel.selectedItem) }
                } catch (e: IOException) {
                    logger.log(Level.SEVERE, e.message, e)
                    Platform.runLater { FinancerExceptionDialog("Financer", IOException("File could not be deleted")) }
                }
            }
        })
    }

    init {
        prepareDialogContent()
        setDialogTitle(get("transaction"))
    }
}